File: backend/database/reminders.py
from datetime import datetime, timezone
from bson import ObjectId
from database.mongodb import get_database


def _serialize(doc):
    return {
        "id": str(doc["_id"]),
        "patient_id": doc["patient_id"],
        "medicine_name": doc["medicine_name"],
        "dosage": doc["dosage"],
        "times": doc["times"],
        "days": doc["days"],
        "is_active": doc.get("is_active", True),
        "created_at": doc.get("created_at", "").isoformat() if isinstance(doc.get("created_at"), datetime) else doc.get("created_at", ""),
    }


def get_reminders(patient_id: str):
    db = get_database()
    docs = list(db.reminders.find({"patient_id": patient_id}))
    return [_serialize(d) for d in docs]


def create_reminder(patient_id: str, data: dict):
    db = get_database()
    data["patient_id"] = patient_id
    data["created_at"] = datetime.now(timezone.utc)
    result = db.reminders.insert_one(data)
    data["_id"] = result.inserted_id
    return _serialize(data)


def update_reminder(reminder_id: str, data: dict):
    db = get_database()
    db.reminders.update_one({"_id": ObjectId(reminder_id)}, {"$set": data})
    doc = db.reminders.find_one({"_id": ObjectId(reminder_id)})
    return _serialize(doc) if doc else None


def delete_reminder(reminder_id: str):
    db = get_database()
    db.reminders.delete_one({"_id": ObjectId(reminder_id)})

File: frontend/src/services/reminders.ts
const API_URL = import.meta.env.VITE_API_URL || 'http://127.0.0.1:8000';

export type ReminderDay = 'Mon' | 'Tue' | 'Wed' | 'Thu' | 'Fri' | 'Sat' | 'Sun';

export const ALL_DAYS: ReminderDay[] = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

export interface Reminder {
  id: string;
  patient_id: string;
  medicine_name: string;
  dosage: string;
  times: string[];
  days: ReminderDay[];
  is_active: boolean;
  created_at: string;
}

export interface ReminderInput {
  medicine_name: string;
  dosage: string;
  times: string[];
  days: ReminderDay[];
  is_active: boolean;
}

async function parseOrThrow(response: Response) {
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.detail || `Request failed (${response.status})`);
  }
  return response.json();
}

export async function getReminders(patientId: string): Promise<Reminder[]> {
  const response = await fetch(`${API_URL}/patients/${encodeURIComponent(patientId)}/reminders`);
  return parseOrThrow(response);
}

export async function createReminder(patientId: string, data: ReminderInput): Promise<Reminder> {
  const response = await fetch(`${API_URL}/patients/${encodeURIComponent(patientId)}/reminders`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  return parseOrThrow(response);
}

export async function updateReminder(
  patientId: string,
  reminderId: string,
  data: Partial<ReminderInput>
): Promise<Reminder> {
  const response = await fetch(
    `${API_URL}/patients/${encodeURIComponent(patientId)}/reminders/${encodeURIComponent(reminderId)}`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    }
  );
  return parseOrThrow(response);
}

export async function deleteReminder(patientId: string, reminderId: string): Promise<void> {
  const response = await fetch(
    `${API_URL}/patients/${encodeURIComponent(patientId)}/reminders/${encodeURIComponent(reminderId)}`,
    { method: 'DELETE' }
  );
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.detail || `Request failed (${response.status})`);
  }
}
